### PR TITLE
enhancement/23567-pattern-anchor-to-point

### PR DESCRIPTION
**Pull request by magdalena-gut**
Added support for anchored pattern fills, see #23567.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/1/222651295669536/project/1204879359430251/task/1211383924433072

### CODE REVIEW

File: ts/Extensions/PatternFill.ts [hunk 1]
```diff
@@ -438,8 +462,32 @@ function onRendererComplexColor(
...
+                            return bbox ?
+                                String(
+                                    Math.round(bbox.x * 100) + '_' +
+                                    Math.round(bbox.y * 100) + '_' +
+                                    Math.round(bbox.width * 100) + '_' +
+                                    Math.round(bbox.height * 100)
+                                ) :
+                                Math.random().toString(36).substr(2, 9);
+                        })();
+                patternId += '-anchored-' + pointIndex;
```
Comment (hubertkozik) [inline ts/Extensions/PatternFill.ts]: If I understand correctly, here we can introduce a bug for multiple series patterns. Now we use only the point index in anchored pattern IDs, so in multi-series charts with identical patterns IDs collide and later series reuse the first point’s pattern (wrong anchor). Consider including series index/uid or element id in the suffix.

File: ts/Extensions/PatternFill.ts [hunk 2]
```diff
@@ -438,8 +462,32 @@ function onRendererComplexColor(
...
+                            const bbox = (element as any).getBBox ?
+                                (element as any).getBBox() : null;
+                            return bbox ?
+                                String(
+                                    Math.round(bbox.x * 100) + '_' +
+                                    Math.round(bbox.y * 100) + '_' +
+                                    Math.round(bbox.width * 100) + '_' +
+                                    Math.round(bbox.height * 100)
+                                ) :
+                                Math.random().toString(36).substr(2, 9);
```
Comment (hubertkozik) [inline ts/Extensions/PatternFill.ts]: I'm not sure if using `Math.random()` is the best approach. Using `Math.random` as a fallback, making IDs non-deterministic and potentially leaking `<pattern>` defs across redraws/exports. Can you check it if maybe a stable key (element id or series+point index)?

File: ts/Extensions/PatternFill.ts [hunk 3]
```diff
@@ -541,8 +572,17 @@ function onRendererComplexColor(
...
+            // point gets a unique ID.
+            if (pattern.anchorToPoint) {
+                pattern = createAnchoredPattern(pattern, element);
+            }
+
             pattern.id = 'highcharts-pattern-' + chartIndex + '-' +
-                hashFromObject(pattern) + hashFromObject(pattern, true);
+                hashFromObject(pattern) + hashFromObject(pattern, true) +
+                (pattern.anchorToPoint ? '-anchored' : '');
         }
```
Comment (hubertkozik) [inline ts/Extensions/PatternFill.ts:586]: `anchorToPoint` currently works only when the pattern ID is auto-generated.

  In `ts/Extensions/PatternFill.ts`, the anchored path (`createAnchoredPattern`) is only executed inside:
  `if (forceHashId || !pattern.id)`

  So when users provide `pattern.id`, points still share one pattern instance, which conflicts with the feature goal ("each column has its own anchored pattern instance").

  Could we also apply the anchored ID path when `pattern.anchorToPoint === true`, even if `pattern.id` is set?

File: samples/unit-tests/color/pattern-fill/demo.js
Comment (hubertkozik) [inline samples/unit-tests/color/pattern-fill/demo.js:1]: Since the OP case is columns (including negative values), can we add a regression test for a column series with mixed positive/negative values and `anchorToPoint: true` to confirm alignment is stable in both directions?

File: samples/highcharts/series/pattern-fill-anchor-to-point/demo.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    }, {
+        name: 'Anchored to Point',
+        data: [3, 8, 13, 18, 1],
+        color: {
+            pattern: {
+                path: 'M 0 0 L 5 5 M 4.5 -0.5 L 5.5 0.5 M -0.5 4.5 L 0.5 5.5',
+                color: '#900000',
+                width: 10,
+                height: 10,
+                anchorToPoint: true // NEW: Each column gets its own pattern
```
Comment (TorsteinHonsi) [inline samples/highcharts/series/pattern-fill-anchor-to-point/demo.js]: Avoid wording like Old and New. This demo will typically not be changed over time, and in some years the feature is not new anymore. Instead, just describe the difference. Also applies to the comment above ("Traditional global pattern").

File: samples/highcharts/series/pattern-fill-anchor-to-point/demo.html [hunk 5]
```diff
@@ -0,0 +1,4 @@
+<script src="https://code.highcharts.com/highcharts.js"></script>
+<script src="https://code.highcharts.com/modules/pattern-fill.js"></script>
+
+<div id="container" style="height: 400px"></div>
```
Comment (TorsteinHonsi) [inline samples/highcharts/series/pattern-fill-anchor-to-point/demo.html]: Looking at det demo, I think it should have a fixed width, or at least max width, so that the outcome is more predictable. Create a `demo.css` file instead of inline style attribute.

File: ts/Extensions/PatternFill.ts [hunk 6]
```diff
@@ -128,6 +128,14 @@ export interface PatternOptionsObject {
     /** @internal */
     _y?: number;
 
+    /**
+     * When true, the pattern is anchored to each individual point rather than
+     * using a global pattern grid. This ensures consistent pattern rendering
+     * across points of different sizes and improves accessibility for narrow
+     * columns. Defaults to false for backward compatibility.
+     */
```
Comment (TorsteinHonsi) [inline ts/Extensions/PatternFill.ts:136]: Add a `@sample` tag pointing to the nice demo

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
Thanks, looks good! Just a couple of minor change requests related to the demo.

**Review by TorsteinHonsi — APPROVED** [review]
